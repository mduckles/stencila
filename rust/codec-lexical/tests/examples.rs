use std::{fs::read_to_string, path::PathBuf};

use codec::{
    common::{eyre::Result, glob::glob},
    format::Format,
    EncodeOptions,
};

use codec_lexical::{decode, encode};
use common_dev::insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};

/// Decode each example of a Lexical document and create JSON and Lexical snapshots
/// including snapshots for losses
#[test]
fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*";

    for path in glob(&format!("{pattern}.lexical"))?
        .chain(glob(&format!("{pattern}.koenig"))?)
        .flatten()
    {
        let format = Format::from_path(&path);
        let ext = format.extension();

        let name = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(&[".", &ext].concat()).map(String::from))
            .expect("should have format suffix");

        let original = read_to_string(path)?;

        let (article, info) = decode(&original, None)?;

        assert_json_snapshot!(format!("{name}.json"), article);
        assert_yaml_snapshot!(format!("{name}.decode.losses"), info.losses);

        let (lexical, info) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format.clone()),
                compact: Some(false),
                ..Default::default()
            }),
        )?;

        assert_snapshot!(format!("{name}.{ext}"), lexical);
        assert_yaml_snapshot!(format!("{name}.encode.losses"), info.losses);

        let (lexical, ..) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format),
                standalone: Some(true),
                compact: Some(false),
                ..Default::default()
            }),
        )?;
        assert_snapshot!(format!("{name}.standalone.{ext}"), lexical);
    }

    Ok(())
}
