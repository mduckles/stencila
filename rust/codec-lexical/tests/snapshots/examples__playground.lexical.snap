---
source: rust/codec-lexical/tests/examples.rs
expression: redact(&lexical)
snapshot_kind: text
---
{
  "root": {
    "type": "root",
    "children": [
      {
        "type": "heading",
        "tag": "h1",
        "children": [
          {
            "type": "text",
            "format": 0,
            "text": "Welcome to the playground"
          }
        ]
      },
      {
        "type": "quote",
        "children": [
          {
            "type": "text",
            "format": 0,
            "text": "In case you were wondering what the black box at the bottom is â€“ it's the debug view, showing the current state of the editor. You can disable it by pressing on the settings control in the bottom-left of your screen and toggling the debug view setting."
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "text",
            "format": 0,
            "text": "The playground is a demo environment built with "
          },
          {
            "type": "text",
            "format": 16,
            "text": "@lexical/react"
          },
          {
            "type": "text",
            "format": 0,
            "text": ". Try typing in "
          },
          {
            "type": "text",
            "format": 1,
            "text": "some text"
          },
          {
            "type": "text",
            "format": 0,
            "text": " with "
          },
          {
            "type": "text",
            "format": 2,
            "text": "different"
          },
          {
            "type": "text",
            "format": 0,
            "text": " formats."
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "text",
            "format": 0,
            "text": "Make sure to check out the various plugins in the toolbar. You can also use "
          },
          {
            "type": "text",
            "format": 0,
            "text": "LOST HashTag"
          },
          {
            "type": "text",
            "format": 0,
            "text": " or @-mentions too!"
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "text",
            "format": 0,
            "text": "If you'd like to find out more about Lexical, you can:"
          }
        ]
      },
      {
        "type": "markdown",
        "markdown": "- Visit the \n\n  [Lexical website](https://lexical.dev/)\n\n   for documentation and more information.\n\n- Check out the code on our \n\n  [GitHub repository](https://github.com/facebook/lexical)\n\n  .\n\n- Playground code can be found \n\n  [here](https://github.com/facebook/lexical/tree/main/packages/lexical-playground)\n\n  .\n\n- Join our \n\n  [Discord Server](https://discord.com/invite/KmG4wQnnD9)\n\n   and chat with the team.\n"
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "text",
            "format": 0,
            "text": "Lastly, we're constantly adding cool new features to this playground. So make sure you check back here when you next get a chance ðŸ™‚."
          }
        ]
      }
    ]
  }
}
